cpp/arcticdb/version/version_map.hpp

```diff
@@ -159,7 +159,10 @@ class VersionMapImpl {
             cached_penultimate_index = ref_entry.keys_[1];
         }
 
-        if (key_exists_in_ref_entry(load_strategy, ref_entry, cached_penultimate_index)) {
+        // The 3 item ref key bypass can be used only when we are loading undeleted versions
+        // because otherwise it might skip versions that are deleted but part of snapshots
+        if (load_strategy.load_objective_ == LoadObjective::UNDELETED_ONLY 
```

Comment:
2 things: Firstly I don't understand the scenario where this is a problem. Is it to do with the slow path where we allow looking for specific version ids that are deleted in the version list but are contained in snapshots?

Secondly, most of the machinery for deciding whether to use the ref-key only or proceed with loading the version list is in `key_exists_in_ref_entry`, wouldn't it be cleaner if this check just returned false from that method?

Reply:
yeah sorry, I should've given more details, this is a minimal repro/test for the scenario:
```
append_small_df(source_lib, "tmp")  # v0
append_small_df(source_lib, "tmp")  # v1
source_lib.snapshot("snap")
source_lib.delete_version("tmp", 1)
append_small_df(source_lib, "tmp")  # v2

library_replicator.process_library(syncing_libraries)
assert compare_libs(source_lib, target_libs)
```
Essentially, version 1 is being deleted but it should be replicated to the secondary because it is in a snapshot.
When using append to write v2, it writes the v2 index key and the v0 index key, because it is the previous undeleted version.
Then when trying to get the index key for v1 during replication, we fail to get it because:

- due to [this optimization of the DOWNTO load strategy](https://github.com/man-group/ArcticDB/blob/master/cpp/arcticdb/version/version_utils.hpp#L341) (as 1 => 0)
- we enter [this logic for loading the index keys](https://github.com/man-group/ArcticDB/blob/master/cpp/arcticdb/version/version_map.hpp#L163), which leads to loading only the index keys in the vref (so only 0 and 2)
- and we fail when we try to filter for v1 in the get_index_key function

As for the second point, yeah I was debating with myself where it would be better to put this check, I will move it to the `key_exists_in_ref_entry` tomorrow